transformer_tools/augmentation.py

```diff
@@ -0,0 +1,367 @@
+        self.de_model = spacy.load(self.pos_model_path)
+        # decides with which probability a word is swapped or it stays the same
+        self.swap_dice = swap_dice
+
+    @functools.lru_cache(maxsize=10_000)
+    @staticmethod
+    def _is_sameword(original_word, new_word):
+        # clean up and lowercase at the same time
+        # only be used at word level
+        def only_word(text_chunk):
+            if len(re.findall(r"[a-z]+", text_chunk.lower())) != 0:
+                return re.findall(r"[a-z]+", text_chunk.lower())[0]
+            return ""
+
+        ori = only_word(original_word.lower())
+        new = only_word(new_word.lower())
+        return ori == new
+
+    @functools.lru_cache(maxsize=10_000)
+    def _gen_spacy_token(self, sent):
+        doc = self.de_model(sent)
+        return list(doc)
+
+    @functools.lru_cache(maxsize=1000)
+    @staticmethod
+    def _is_validword(spacy_token, valid_pos=None):
+        """TODO: fix docstring.
+
+        :param spacy_token: spacy token of the word
+        :param valid_pos:
+        :return: Boolean: True / False
+        """
+        valid_pos = ["VERB", "ADV", "NOUN", "ADJ"] if valid_pos is None else valid_pos
+        return spacy_token.pos_ in valid_pos
+
+    def _swap_with_weights(self, ori_word, prob):
+        # swap a word with candidates or stay the same depending on given probability
+        # prob: probability of being swapped
+        swap = self.get_candidates(ori_word)
+        return random.choices(population=[ori_word, random.choice(swap)], weights=[1 - prob, prob])
+
+    def get_candidates(self, word):
+        """TODO: add docstring."""
+        raise NotImplementedError
+
+    def _generate(self, sent):
+        # candidate is based on part of speech filtering with spacy pos tag
+        # candidate_generation is applied on every sent (one instance)
+        # generate one augmentation
+        # aug_text is initialized as list of tokens from the input sent
+        aug_text = [token.text for token in self._gen_spacy_token(sent)]
+        # print("tokens: ", aug_text)
+        # for token in self._gen_spacy_token(sent):
+        token_list = self._gen_spacy_token(sent)
+        valid_token_idx = [idx for (idx, tok) in enumerate(token_list) if self._is_validword(tok)]
+        # print("valid tokens: ", [(idx, token_list[idx].text) for idx in valid_token_idx])
+        # select the token to be swapped
+        if len(valid_token_idx) != 0:
+            # print("selected_index", selected_index)
+            # then we swap the selected...
+            aug_text = [
+                self._swap_with_weights(aug_text[idx], self.swap_dice)
+                if idx in valid_token_idx
+                else word.lower()
+                for (idx, word) in enumerate(aug_text)
+            ]
+        out = " ".join(aug_text).strip()
+        out = re.sub(" +", " ", out)
+        out = re.sub(" ,", ",", out)
+        out = re.sub(r" \.", ".", out)
```

Comment:
@sitongye instead of this we should be able to use the `whitespace_` attribute of the spacy token.
That would be a cleaner way to concatenate the text.
See: https://spacy.io/api/token#attributes

Reply:
do you think it's worth it if I transform the fasttext embedding to be in the spacy format so that we can simply swap the tokens? https://spacy.io/usage/linguistic-features#vectors-similarity

Reply:
Not sure. Maybe we should write an abstraction to change the way how we swap token?